charmcraft.yaml

```diff
@@ -0,0 +1,132 @@
+# This file configures Charmcraft.
+# See https://juju.is/docs/sdk/charmcraft-config for guidance.
+
+# (Required)
+# The charm package name, no spaces
+# See https://juju.is/docs/sdk/naming#heading--naming-charms for guidance.
+name: calibre-web
+
+
+# (Required)
+# The charm type, either 'charm' or 'bundle'.
+type: charm
+
+
+# (Recommended)
+title: Charm Template
+
+
+# (Required)
+summary: A very short one-line summary of the charm.
+
+
+# (Required)
+description: |
+  A single sentence that says what the charm is, concisely and memorably.
+
+  A paragraph of one to three short sentences, that describe what the charm does.
+
+  A third paragraph that explains what need the charm meets.
+
+  Finally, a paragraph that describes whom the charm is useful for.
+
+
+# (Required for 'charm' type)
+# A list of environments (OS version and architecture) where charms must be
+# built on and run on.
+bases:
+  - build-on:
+    - name: ubuntu
+      channel: "22.04"
+    run-on:
+    - name: ubuntu
+      channel: "22.04"
+
+
+# (Optional) Configuration options for the charm
+# This config section defines charm config options, and populates the Configure
+# tab on Charmhub.
+# More information on this section at https://juju.is/docs/sdk/charmcraft-yaml#heading--config
+# General configuration documentation: https://juju.is/docs/sdk/config
+config:
+  options:
+    library-write:
+      default: skip
+      type: string
+      description: |
+        What to do when /books/ contains something already. Options are:
+        skip (do nothing),
+        clean (delete the contents of /books/ and then extract the provided library or the default if none is provided).
+        Changing this value after storage has already been attached will immediately re-run the storage attached event handler,
+        potentially resulting in cleaning the current contents of /books/.
+
+# removed library-write option:
+        #overwrite (extract the provided library on top of the existing one if one is provided),
+
+# The containers and resources metadata apply to Kubernetes charms only.
+# See https://juju.is/docs/sdk/metadata-reference for a checklist and guidance.
+# Remove them if not required.
+
+
+# Your workload’s containers.
+containers:
+  calibre-web:
+    resource: calibre-web-image
+    mounts:
+      - storage: books
+        location: /books
+
+storage:
+  books:
+    type: filesystem
+    description: Storage mount for Calibre Library
+    location: /books
+    minimum-size: 1G
+    properties:
+      - transient
+
+parts:
+  charm:
+    plugin: charm
+    source: .
+  default-calibre-library:
```

Comment:
I know that there have been changes with this in recent versions of charmcraft, but I thought that the charm plug-in would pull in all the files by default, except for whatever is in `.jujuignore`? I might be wrong or that might be specific to some charmcraft version, though.

Reply:
By default, with no `parts`, I think it works as you say.

Once you add a part, in my case `default-calibre-library`, it will only pack that part, skipping *everything else* (`src/charm.py`, etc), which was a surprise for me. I can't find it now, but I seem to recall one piece of documentation (not [this one](https://juju.is/docs/sdk/charmcraft-yaml#heading--parts)) mentioning that after some juju (or charmcraft?) version, you'd have to provide the `charm` plugin explicitly. The documentation I looked at just now doesn't seem to mention this.

One thing I didn't figure out was why `charmcraft pack` now seems to pull in every file in my working directory, rather than just the known charm files and the ones specified in `parts`, which is how I thought it would work. I guess it all has to be set up in `.jujuignore`?

Reply:
Yeah, I'm really unsure on what the situation is on this at the moment (and when it changed, etc). I don't think the docs do a great job of explaining this, although I know they are working on much more elaborate charmcraft docs, so hopefully that will solve it.